### ISSUE TITLE
q-btn violate HTML5 spec

### ISSUE THREAD

**Issue by kdmon** [quasarframework/quasar#5888]
Hi

The q-btn rendered markup does not conform to the HTML5 spec. This may be detrimental to achieving a high level of accessibility (see https://dequeuniversity.com/tips/use-valid-markup)

The basic rendered q-btn markup looks like this:

```html
<button tabindex="0" type="button" class="...">
  <div tabindex="-1" class="q-focus-helper"></div>
  <div class="q-btn__wrapper col row q-anchor--skip">
    <div class="..."><div>
       Button text
     </div>
   </div>
  </div>
</button>
```

Testing in the w3.org validator results in two types of validation errors:

1. "Element div not allowed as child element of button."
2. "An element with the attribute tabindex must not appear as a descendant of the button element"

Firstly, the button tag contains div tags (categorized in the spec as "flow content") which are not permitted by the HTML5 spec. Only "phrasing content" (e.g. span) is allowed inside button tags.

Secondly, tabindex is not permitted (as they indicate "interactive content" which is also not allowed inside a button element).

The remedy is to remove divs and tabindex inside of button elements. Below you can see how Vuetify.js button markup passes validation by using span elements and avoiding  tabindex:

```html
<button type="button" class="...">
  <span class="v-btn__content">
  <span>Click Me</span></span>
</button>
```

Here is the minimal reproducible example I used to test on https://validator.w3.org/nu/#textarea

```html
<!DOCTYPE html>
<html lang="en">
<head>
<title>q-btn example</title>
</head>
<body>
<button tabindex="0" type="button" class="...">
  <div tabindex="-1" class="q-focus-helper"></div>
  <div class="q-btn__wrapper col row q-anchor--skip">
    <div class="..."><div>
       Button text
     </div>
   </div>
  </div>
</button>
</body>
</html>
```

Finally, the HTML button spec is here: https://html.spec.whatwg.org/multipage/form-elements.html#the-button-element

Thanks!

**Comment by pdanpdan** [comment]
You do realize the validator is very good when a negative tabindex is mistaken for a non negative one :)

**Comment by masi** [comment]
The button code looks really odd. I haven't seen any a11y articles that suggest suggest a construct. I hope that it has been tested with a variety if screen readers to make sure it does any good. My gut feeling says the button should be placed inside the wrapper div.

**Comment by kdmon** [comment]
Any news on this issue? Our product team is considering migrating to Vuetify due to uncertainty about screen reader compatibility. How difficult would it be to patch this?

**Comment by pdanpdan** [comment]
Did you have any problem with the screen reader?

**Comment by masi** [comment]
Here are the relevant parts of the spec which forbid such a markup:

"Content model:
Phrasing content, but there must be no interactive content descendant."

https://www.w3.org/TR/html52/sec-forms.html#the-button-element
https://www.w3.org/TR/html52/dom.html#phrasing-content-2

"The tabindex attribute can also make any element into interactive content."

https://www.w3.org/TR/html52/dom.html#interactive-content-2

For accessibility reasons the DIVs are no problem at all. Just using SPANs with another CSS display-style will make the validator happy.

What's the point of the nested tabindex or actually of the helper? If the focus-helper would either wrap the button or be placed before the button then there wouldn't be any violation of the spec.

**Comment by kdmon** [comment]
Hi and thanks for your fast response. I apologise if I came across as demanding stuff, it was not my intention (written in haste). I'm a fan of quasar, appreciate your hard work and want to help make it ever better!

A quick test suggests that keeping the q-btn component structure, but changing the child divs to spans and not setting a tabindex=-1 on e.g. q-focus-helper will make the markup conform to spec. I have not seen any obvious negative impact on functionality, but I'm unsure if the above changes would have any side effects that I am unaware of. To echo Masi's question, what is the reason for setting tabindex=-1 on the q-focus-helper? (everything seems to work fine without this).

If there are no detrimental impact, the above would be a quick fix and I'm happy to submit a PR for review.

Thanks again

**Comment by pdanpdan** [comment]
@kdmon Sure, please make the PR - it helps us, thank you. But please test it also on SSR and mobiles (especially iOS web and cordova/capacitor) - check to have no hydration errors.

**Comment by rstoenescu** [comment]
Fix will be available in "quasar" v1.11.0
